dali/imgcodec/decoders/opencv_fallback_test.cc

```diff
@@ -0,0 +1,151 @@
+  DecodeParams params;
+  int64_t n = volume(info.shape);
+  ASSERT_GE(n, 0);
+  ASSERT_LE(n, 100000000);  // sanity check - less than 100M elements
+  auto mem = mm::alloc_raw_unique<OutputType, mm::memory_kind::host>(n);
+  SampleView<CPUBackend> sv(mem.get(), info.shape, type2id<OutputType>::value);
+  auto result = instance->Decode(sv, &source, params);
+  if (result.exception) {
+    EXPECT_NO_THROW(std::rethrow_exception(result.exception));
+  }
+  ASSERT_TRUE(result.success);
+
+  cv::Mat m = cv::imread(fname, cv::IMREAD_COLOR | cv::IMREAD_IGNORE_ORIENTATION);
+  cv::cvtColor(m, m, cv::COLOR_BGR2RGB);
+  int64_t out_row_stride = info.shape[1] * info.shape[2];
+  for (int y = 0; y < m.rows; y++) {
+    const OutputType *out_row = sv.data<OutputType>() + y * out_row_stride;
+    const uint8_t *ref_row = m.ptr<uint8_t>(y);
+    for (int x = 0; x < m.cols; x++) {
+      for (int c = 0; c < 3; c++) {
+        ASSERT_EQ(out_row[3*x + c], ConvertSatNorm<OutputType>(ref_row[3*x + c]))
+          << " at " << x << ", " << y << ", " << c;
+      }
+    }
+  }
+}
+
+TYPED_TEST(OpenCVFallbackDecodeTest, DecodeROI) {
+  using OutputType = TypeParam;
+
+  ThreadPool tp(4, CPU_ONLY_DEVICE_ID, false, "OpenCV decoder test");
+  OpenCVDecoder decoder;
+  auto fname = img_palette;
+  auto source = ImageSource::FromFilename(fname);
+  auto instance = decoder.Create(CPU_ONLY_DEVICE_ID, tp);
+  ASSERT_NE(instance, nullptr);
+
+  BmpParser parser;
+  EXPECT_TRUE(parser.CanParse(&source));
+  ImageInfo info = parser.GetInfo(&source);
+
+  DecodeParams params;
+  params.use_roi = true;
+  params.roi.begin = { 13, 17 };
+  params.roi.end = { info.shape[0] - 23, info.shape[1] - 19 };
+  TensorShape<> out_shape;
+  out_shape = shape_cat(params.roi.shape(), 3);
+  int64_t n = volume(out_shape);
+  ASSERT_GE(n, 0);
+  ASSERT_LE(n, 100000000);  // sanity check - less than 100M elements
+  auto mem = mm::alloc_raw_unique<OutputType, mm::memory_kind::host>(n);
+  SampleView<CPUBackend> sv(mem.get(), out_shape, type2id<OutputType>::value);
+  auto result = instance->Decode(sv, &source, params);
+  if (result.exception) {
+    EXPECT_NO_THROW(std::rethrow_exception(result.exception));
+  }
+  ASSERT_TRUE(result.success);
+
+  cv::Mat m = cv::imread(fname, cv::IMREAD_COLOR | cv::IMREAD_IGNORE_ORIENTATION);
+  cv::Rect roi_rect(cv::Point2i(params.roi.begin[1], params.roi.begin[0]),
+                    cv::Point2i(params.roi.end[1],   params.roi.end[0]));
+  m = m(roi_rect);
+  cv::cvtColor(m, m, cv::COLOR_BGR2RGB);
+  int64_t out_row_stride = out_shape[1] * out_shape[2];
+  for (int y = 0; y < m.rows; y++) {
+    const OutputType *out_row = sv.data<OutputType>() + y * out_row_stride;
+    const uint8_t *ref_row = m.ptr<uint8_t>(y);
+    for (int x = 0; x < m.cols; x++) {
+      for (int c = 0; c < 3; c++) {
+        ASSERT_EQ(out_row[3*x + c], ConvertSatNorm<OutputType>(ref_row[3*x + c]))
+          << " at " << x << ", " << y << ", " << c;
+      }
+    }
+  }
```

Comment:
Extract this as a function with optional ROI and reuse? I can imagine we will be writing more such tests for other image formats, and we don't want to write this each time

Reply:
Even the whole test could be parametrized with the filename (just a function taking a filename that is reused for other image format tests)

Reply:
I'm not so sure. This test is exceptional, because we use OpenCV internally, so a direct comparison against OpenCV works. For other formats it won't work quite as well - at least we should be ready for handling different bit depths and having the reference file in a different format than the original encoded one... How about doing it in a follow-up?

Reply:
OK, but still you are copying a bunch of code twice at least (and I can see more usage in the future). The function still makes sense to me